### PR TITLE
feat: made spent time format configurable

### LINKED ISSUE #554: Add option to show playtime in hours instead of days (or even custom format)
### Role

I really like being able to see my playtime of each instance (and the total playtime), however I don't like seeing the playtime in terms of days

### Suggestion

I would like to have the option (since I understand that people prefer the current setting) to show the playtime purely in hours instead of days.

### Benefit

So that I can more easily understand/gauge the playtime and compare.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

To further expand, here is an example. Let's say my playtime on an instance is `totalTimePlayed=275193` (from `instance.cfg`, time is stored in seconds).

Current UI shows my total play time as "3d 4h 26m" (sidenote, the units are incorrect, see #553).

I suggest that we could show the playtime as "76h 26min" which is a common way of showing playtime in most game launchers.

This setting would be located in Settings > Launcher > User Interface.

Ideally, there could be a third option to add a text input where the user can input a formatting string to customize the display however they want. Here's a suggestion of what codes could be available and what they would do: 
- Days:
  - `%d%`: full amount of days (rounded down). Example: "3"
  - `%dd%`: full amount of days, with 1 digit of precision. Example: "3.2"
  - `%ddd%`: full amount of days, with 2 digit of precision. Example: "3.18"
- Hours:
  - `%h%`: amount of hours (full days deducted) (rounded down). Example: "4"
  - `%hh%`: amount of hours (full days deducted), with 1 digit of precision. Example: "4.4"
  - `%hhh%`: amount of hours (full days deducted), with 2 digit of precision. Example: "4.43"
  - `%fh%`: full amount of hours (rounded down). Example: "76"
  - `%fhh%`: full amount of hours, with 1 digit of precision. Example: "76.4"
  - `%fhhh%`: full amount of hours, with 2 digit of precision. Example: "76.43"
- Minutes:
  - `%m%`: amount of minutes (full days and full hours deducted) Example: "26"
  - `%fm%`: full amount of minutes. Example: "4586"

And here's a bunch of examples for a playtime of 275193s:

| **Time string**                  | **Output**              | **Notes**                    |
|----------------------------------|-------------------------|------------------------------|
| `"%d%d %h%h %m%min"`             | 3d 4h 26min             | Current format               |
| `"%fh%h %m%min"`                 | 76h 26min               | Suggested alternative format |
| `"%fhh% hours"`                  | 76.4 hours              |                              |
| `"%d% days and %h% hours"`       | 3 days and 4 hours      |                              |
| `"%fm% minutes"`                 | 4,586 minutes           |                              |
| `"%d% days %h% hours (%fhhh%h)"` | 3 days 4 hours (76.43h) |                              |

From what I can see, this formatting is handled by [`MMCTime.cpp`](https://github.com/PrismLauncher/PrismLauncher/blob/a9d4370ad462b4ad3dd91f6bc38b40262967b5b2/launcher/MMCTime.cpp)

Note: The closest issue I could find is #216 but I don't think this covers the same topic.

### PR DESCRIPTION
**Pull request by Trial97**
fixes #554

Small changes that differ from the requirement:

- Days:
    - %d full amount of days (rounded down). Example: "3"
    - %d.1: full amount of days, with 1 digit of precision. Example: "3.2"
    - %d.2: full amount of days, with 2 digits of precision. Example: "3.18"
- Hours:
    - %h: amount of hours (full days deducted) (rounded down). Example: "4"
    - %h.1: amount of hours (full days deducted), with 1 digit of precision. Example: "4.4"
    - %h.2: amount of hours (full days deducted), with 2 digits of precision. Example: "4.43"
    - %H: full amount of hours (rounded down). Example: "76"
    - %H.1: full amount of hours, with 1 digit of precision. Example: "76.4"
    - %H.2: full amount of hours, with 2 digits of precision. Example: "76.43"
- Minutes:
    - %m: amount of minutes (full days and full hours deducted) Example: "26"
    - %M: full amount of minutes. Example: "4586"
- Seconds:
    - %s: amount of seconds (full days, full hours deducted,and full minutes deducted) Example: "26"
    - %S: full amount of seconds. Example: "4586"

The precision can be set exactly by suffixing the option with a dot and the wanted precision, e.g. %d.4 => days with digits of precision.

### CODE REVIEW

File: launcher/MMCTime.cpp [hunk 1]
```diff
@@ -78,25 +78,124 @@ QString Time::humanReadableDuration(double duration, int precision) {
...
+            inOption = false;
+            int precision = 0;
+            s.flush();
+            if (precSeg.startsWith('.')) {
+                precision = precSeg.remove(0, 1).toInt();
+            }
+            s.reset();
+            precSeg = "";
+            double dc = 0;
+            if (option == 'd') {
```
Comment (getchoo) [inline launcher/MMCTime.cpp]: hate to be that person but...couldn't this be a `switch` statement? i think it'd just look a little nicer :p

**Review by getchoo — CHANGES_REQUESTED** [review]
i really like this! a few suggestions though:
- validate `GameTimeFormat`
  - this prevents issues where a user may have input improper or invalid formatting and having weird things appear in the ui. i think it would be best to revert back to the defaults in this case
- maybe ignore the `GameTimeFormat` when it doesn't apply to the given instance
  - in cases where i only specify `%d%d`, the `total played: ` line has no characters following it. i think it would be best to revert back to the default here as well or try to convert the playtime to that format; another solution could be to allow for per instance configuration. imo reverting to the default or converting to the measurement specified by the formatting string would be the best option here, while still being pretty simple
- simply the syntax
  - similar to `date` in linux, i think just specifying `%d` for day, `%h` for hour, `%m` for minute, and `%s` for seconds would make more sense. using the same character twice in most cases (while also having two different formatting strings only for minutes) is a little odd to me and i don't believe it's something users would do on purpose a lot of the time

File: launcher/MMCTime.cpp (hunk 1 above)
Comment (Trial97) [reply launcher/MMCTime.cpp]: changed most of the code there to use switch

**Comment by Trial97** [comment]
Made most of the suggested changes. Here are some small notes:
- I kept the extra options(%H,%M, and %S) as they are used when you want to display the duration in that respective unit only, they are meant to be used together with precision (also they are a requirement from the issue)
- this format should be similar to the date in linux with the additions of the extra options and controllable precision
- now if the format is empty it fallbacks to the default format
- if the result is empty because of the trimmed zero values the function is recalled but keeps the zero everywhere(https://github.com/PrismLauncher/PrismLauncher/blob/b9b51953d2ef211dd28177d8252bdccb39cae4a5/tests/Time_test.cpp#L46)
- I did not do validation as I think is unnecessary, if the format is not empty and doesn't contain any known options the text will be printed as is, missing only the group of options that are not valid( the character '%' and the next one) https://github.com/PrismLauncher/PrismLauncher/blob/b9b51953d2ef211dd28177d8252bdccb39cae4a5/tests/Time_test.cpp#L57-L60

**Comment by Scrumplex** [comment]
I don't think we need this much granularity for this, though? I think a simple checkbox to show the playtime in hours is fine.

**Comment by Trial97** [comment]
fair: I will close this after I open the  other one

**Comment by Trial97** [comment]
closing this in favor of:https://github.com/PrismLauncher/PrismLauncher/pull/1536

**Comment by victorleblais** [comment]
I still think this would be neat as an advanced/optional feature that could be hidden behind a dropdown (with three options: days/hours, just hours, or "advanced", which shows the text input below the dropdown to type a custom format).

I always like having more options, but it's fine if they're hidden behind more clicks in order to keep the "basic behaviors" quick and accessible.

**Comment by Scrumplex** [comment]
> I still think this would be neat as an advanced/optional feature that could be hidden behind a dropdown (with three options: days/hours, just hours, or "advanced", which shows the text input below the dropdown to type a custom format).
> 
> I always like having more options, but it's fine if they're hidden behind more clicks in order to keep the "basic behaviors" quick and accessible.

Advanced users can just go and extract the play time in seconds from the `instance.cfg` file and do whatever they want with that.

**Pull request closed without merging**